lib/pbench/server/cache_manager.py

```diff
@@ -515,25 +535,62 @@ def extract(tarball_path: Path, path: Path) -> Inventory:
 
         Returns:
             An inventory object that mimics an IO[bytes] object while also
-            maintaining a reference to the tarfile TarFile object to be
-            closed later.
+            maintaining a reference to the subprocess.Popen object to be
+            cleaned up later.
 
         Raise:
-            TarballNotFound on failure opening the tarball
             CacheExtractBadPath if the target cannot be extracted
+            TarballUnpackError on other tar-command failures
+            Any exception raised by subprocess.Popen()
         """
-        try:
-            tar = tarfile.open(tarball_path, "r:*")
-        except Exception as exc:
-            raise TarballNotFound(str(tarball_path)) from exc
-        try:
-            stream = tar.extractfile(str(path))
-        except Exception as exc:
-            raise CacheExtractBadPath(tarball_path, path) from exc
-        else:
-            if not stream:
-                raise CacheExtractBadPath(tarball_path, path)
-            return Inventory(stream, tar)
+        tar_path = shutil.which("tar")
+        if tar_path is None:
+            raise TarballUnpackError(
+                tarball_path, "External 'tar' executable not found"
+            )
+
+        # The external tar utility offers better capabilities than the
+        # Standard Library package, so run it in a subprocess:  extract
+        # the target member from the specified tar archive and direct it to
+        # stdout; we expect only one occurrence of the target member, so stop
+        # processing as soon as we find it instead of looking for additional
+        # instances of it later in the archive -- this is a huge savings when
+        # the archive is very large.
+        tarproc = subprocess.Popen(
+            [str(tar_path), "xOf", tarball_path, "--occurrence=1", path],
+            stdin=subprocess.DEVNULL,
+            stdout=subprocess.PIPE,
+            stderr=subprocess.PIPE,
+        )
+
+        # Wait for one of two things to happen:  either the subprocess produces
+        # some output or it exits.
+        while not tarproc.stdout.peek() and tarproc.poll() is None:
+            sleep(0.02)
+
+        # If the return code is None (meaning the command is still running) or
+        # is zero (meaning it completed successfully), then return the stream
+        # containing the extracted file to our caller, and return the Popen
+        # object to ensure it stays "alive" until our caller is done with the
+        # stream.
+        if not tarproc.returncode:
+            return Inventory(tarproc.stdout, tarproc)
+
+        # The tar command was invoked successfully (otherwise, the Popen()
+        # constructor would have raised an exception), but it exited with
+        # an error code.  We have to glean what went wrong by looking at
+        # stderr, which is fragile but the only option.  Rather than
```

Comment:
Yeah, too bad `tar` doesn't have useful (or even well defined) non-zero return codes. (Sigh.)

Reply:
Actually, the error text (at least the ones that we're interested in) come from `strerror()` which, by this point, is pretty stable and standard...so we _could_ consider recognizing the messages...but I'm satisfied with the current arrangement.